source/Calamari/Kubernetes/Authentication/AwsCliAuth.cs

```diff
@@ -0,0 +1,154 @@
+                return;
+
+            log.Verbose("Attempting to authenticate with aws-iam-authenticator");
+            SetKubeConfigAuthenticationToAwsIAm(user, clusterName);
+        }
+
+        bool TrySetKubeConfigAuthenticationToAwsCli(string clusterName, string clusterUrl, string user)
+        {
+            if (!awsCli.TrySetAws())
+            {
+                log.Verbose("Could not find the aws cli, falling back to the aws-iam-authenticator.");
+                return false;
+            }
+
+            ConfigureAwsCli();
+
+            try
+            {
+                var awsCliVersion = awsCli.GetAwsCliVersion();
+                var minimumAwsCliVersionForAuth = new SemanticVersion("1.16.156");
+                if (awsCliVersion.CompareTo(minimumAwsCliVersionForAuth) > 0)
+                {
+                    var region = GetEksClusterRegion(clusterUrl);
+                    if (!string.IsNullOrWhiteSpace(region))
+                    {
+                        var apiVersion = awsCli.GetEksClusterApiVersion(clusterName, region);
+                        SetKubeConfigAuthenticationToAwsCli(user, clusterName, region, apiVersion);
+                        return true;
+                    }
+
+                    log.Verbose("The EKS cluster Url specified should contain a valid aws region name");
+                }
+
+                log.Verbose(
+                    $"aws cli version: {awsCliVersion} does not support the \"aws eks get-token\" command. Please update to a version later than 1.16.156");
+            }
+            catch (Exception e)
+            {
+                log.Verbose(
+                    $"Unable to authenticate to {clusterUrl} using the aws cli. Failed with error message: {e.Message}");
+            }
+
+            return false;
+        }
+
+        void SetKubeConfigAuthenticationToAwsIAm(string user, string clusterName)
+        {
+            var kubectlVersion = kubectl.GetVersion();
+            var apiVersion = kubectlVersion.Some() && kubectlVersion.Value > new SemanticVersion("1.23.6")
+                ? "client.authentication.k8s.io/v1beta1"
+                : "client.authentication.k8s.io/v1alpha1";
+
+            kubectl.ExecuteCommandAndAssertSuccess(
+                "config",
+                "set-credentials",
+                user,
+                "--exec-command=aws-iam-authenticator",
+                $"--exec-api-version={apiVersion}",
+                "--exec-arg=token",
+                "--exec-arg=-i",
+                $"--exec-arg={clusterName}");
+        }
+
+        void ConfigureAwsCli()
+        {
+            if (!environmentVars.ContainsKey("AWS_ACCESS_KEY_ID"))
+            {
+                var awsEnvironmentGeneration =
+                    AwsEnvironmentGeneration.Create(log, deploymentVariables).GetAwaiter().GetResult();
+                environmentVars.AddRange(awsEnvironmentGeneration.EnvironmentVars);
+            }
```

Comment:
If we are running a kubectl script step, this has already been done so we can skip it the second time.

Reply:
To clarify, what's the scenario where this path gets invoked twice?

Reply:
Discussed over zoom, but for posterity - When we run a kubectl script step, calamari will handle the authentication via the `AwsScriptWrapper` (I think, haven't double checked the exact location) and then again when we authenticate for kubectl. This doesn't get run for a step package framework step.